### PR TITLE
 feat(pm_list): add required fields for giropay 

### PR DESCRIPTION
**Pull request by AkshayaFoiger**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Resolves [#3622](https://github.com/juspay/hyperswitch/issues/3622)
### Test Cases
1. Create a giropay payment with mollie with empty payment method object. The payment should succeed. 

```
{
  "amount": 1000,
  "currency": "EUR",
  "confirm": true,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "amount_to_capture": 1000,
  "customer_id": "StripeCustomer",
  "email": "abcdef123@gmail.com",
  "name": "John Doe",
  "phone": "999999999",
  "phone_country_code": "+65",
  "description": "Its my first payment request",
  "authentication_type": "three_ds",
  "return_url": "https://google.com",
  "statement_descriptor_name": "joseph",
  "statement_descriptor_suffix": "JS",
  "payment_method": "bank_redirect",
  "payment_method_type": "giropay",
  "payment_method_data": {
    "bank_redirect": {
      "giropay": {
      }
    }
  }
}
```

Must be tested for each of this connectors

1.ACI
```
Required fields: 
    payment_method_data.bank_redirect.giropay.country
```

2. Adyen
```
None
```
3. GlobalPayments
```
Required fields: 
    billing.address.country
```
4. Mollie
```
None
```
5. Nuvie
```
Required fields: 
     email
     billing.address.first_name
     billing.address.last_name
     billing.address.country
```
6. Paypal 
```
Required fields: 
     payment_method_data.bank_redirect.giropay.billing_details.billing_name
     payment_method_data.bank_redirect.giropay.country
```
7. Shift4
```
None
```
8. Stripe  
```
Required fields 
     payment_method_data.bank_redirect.giropay.billing_details.billing_name
```
9. Trustpay
```
Required fields: 
     billing.address.first_name
     billing.address.line1
     billing.address.city
     billing.address.zip
     billing.address.country
```

-> Create a payment with confirm `false`

-> list payment methods, with client secret and publishable key

```
curl --location 'http://localhost:8080/account/payment_methods?client_secret={{client_secret}}' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key:{{api_key}}' \
--data ''
```
 Required fields must be populated, according to the connector

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -1258,15 +1258,15 @@ pub enum BankRedirectData {
...
 
         /// The country for bank payment
         #[schema(value_type = CountryAlpha2, example = "US")]
-        country: api_enums::CountryAlpha2,
+        country: Option<api_enums::CountryAlpha2>,
 
         /// The preferred language
         #[schema(example = "en")]
-        preferred_language: String,
+        preferred_language: Option<String>,
```
Comment (prasunna09) [inline crates/api_models/src/payments.rs:1269]: why do we have changes for Sofort pm type in this pr? I observed there was another PR for sofort, please undo these changes if not necessary

Comment (AkshayaFoiger) [reply crates/api_models/src/payments.rs]: Yes, there is another separate PR for Sofort. Sofort changes appeared in this branch as  `dynamic_field/giropay` branch is built upon `dynamic_field/sofort` branch whilst the PR pointed to main. I have changed the base branch. Can you please review it, again.

File: crates/router/src/connector/worldline/transformers.rs [hunk 2]
```diff
@@ -383,11 +383,16 @@ fn make_bank_redirect_request(
...
                 PaymentMethodSpecificData::PaymentProduct816SpecificInput(Box::new(Giropay {
                     bank_account_iban: BankAccountIban {
-                        account_holder_name: billing_details.billing_name.clone().ok_or(
-                            errors::ConnectorError::MissingRequiredField {
+                        account_holder_name: billing_details
+                            .clone()
+                            .ok_or(errors::ConnectorError::MissingRequiredField {
+                                field_name: "giropay.billing_details",
+                            })?
+                            .billing_name
```
Comment (prasunna09) [inline crates/router/src/connector/worldline/transformers.rs:391]: please use the utils method - `get_billing_name()` here

**Review by prasunna09 — CHANGES_REQUESTED** [review]

File: crates/router/src/configs/defaults.rs [hunk 3]
```diff
@@ -4465,14 +4465,244 @@ impl Default for super::settings::RequiredFields {
...
+                                    }
+                                ),
+                                (
+                                    enums::Connector::Adyen,
+                                    RequiredFieldFinal {
+                                        mandate: HashMap::new(),
+                                        non_mandate: HashMap::new(),
+                                        common: HashMap::new(),
+                                    }
+                                ),
```
Comment (prasunna09) [inline crates/router/src/configs/defaults.rs:4514]: if Adyen supports mandates, could you please confirm if this config is present in toml file?
```
[mandates.supported_payment_methods]
bank_redirect.giropay
```
Could you please check for other connectors?

**Review by prasunna09 — CHANGES_REQUESTED** [review]

File: crates/router/src/configs/defaults.rs (hunk 3 above)
Comment (AkshayaFoiger) [reply crates/router/src/configs/defaults.rs:4496]: Adyen supports mandates. I will add it to the config files

Comment (pixincreate) [reply crates/router/src/configs/defaults.rs:4496]: Also, make sure you update the env files which is located at https://github.com/juspay/hyperswitch/tree/main/config/deployments

**Review by prasunna09 — CHANGES_REQUESTED** [review]

File: crates/router/src/configs/defaults.rs [hunk 4]
```diff
@@ -4467,14 +4467,226 @@ impl Default for super::settings::RequiredFields {
...
+                                (
+                                    enums::Connector::Aci,
+                                    RequiredFieldFinal {
+                                        mandate: HashMap::new(),
+                                        non_mandate: HashMap::from([
+                                            (
+                                                "payment_method_data.bank_redirect.giropay.country".to_string(),
+                                                RequiredFieldInfo {
+                                                    required_field: "payment_method_data.bank_redirect.giropay.country".to_string(),
+                                                    display_name: "bank_account_iban".to_string(),
```
[the remark this answers is not in the document]
Comment (AkshayaFoiger) [reply crates/router/src/configs/defaults.rs]: It was a typo. Changed it

**Review by prasunna09 — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]
